### PR TITLE
rgw: address crash and race in RGWIndexCompletionManager

### PR DESCRIPTION
**Pull request by ivancich**
An atomic int was used in a modulo operator to distribute contention
among a set of locks and to track completions. Because it was an int,
enough increments would cause it to go negative (due to
twos-complement encoding and overflow) thereby causing a
crash. Additionally, even though it was atomic, the read and increment
were separate operations, leading to a race.

This commit addresses both of these issues.

Signed-off-by: J. Eric Ivancich <ivancich@redhat.com>

Fixes: https://tracker.ceph.com/issues/55131

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
@ivancich which downstream branches do we need this fix on?

**Review by adamemerson — APPROVED** [review]

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -774,7 +774,7 @@ struct complete_op_data {
 
 class RGWIndexCompletionManager {
   RGWRados* const store;
-  const int num_shards;
+  const uint num_shards;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:777]: would prefer uint32_t or at worst, "unsigned int"

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -784,7 +784,10 @@ class RGWIndexCompletionManager {
   bool _stop{false};
   std::thread retry_thread;
 
-  std::atomic<int> cur_shard {0};
+  // used to distribute the completions and the locks they use across
+  // their respective vectors; it will get incremented and can wrap
+  // around back to 0 without issue
+  std::atomic<uint> cur_shard {0};
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:790]: ditto

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -806,10 +809,8 @@ class RGWIndexCompletionManager {
     completions.clear();
   }
   
-  int next_shard() {
-    int result = cur_shard % num_shards;
-    cur_shard++;
-    return result;
+  uint next_shard() {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:812]: yup

**Review by mattbenjamin — APPROVED** [review]
could you hack the int decl?

**Comment by ivancich** [comment]
Teuthology Results:

- https://pulpito.ceph.com/ivancich-2022-04-29_21:08:04-rgw-wip-eric-testing-1-distro-basic-smithi/
- https://pulpito.ceph.com/ivancich-2022-04-29_21:07:42-rgw-wip-eric-testing-1-distro-basic-smithi/

**Pull request merged**